File: api/schemas/create_organization/create_organization.py
import graphene

from graphql import GraphQLError

from app import logger
from db import db_session
from functions.auth_wrappers import require_token
from functions.auth_functions import is_super_admin
from functions.input_validators import cleanse_input
from functions.slugify import slugify_value
from models import Organizations
from scalars.organization_acronym import Acronym


class CreateOrganization(graphene.Mutation):
    """
    Mutation allows the creation of an organization inside the database.
    """

    class Arguments:
        acronym = Acronym(description="Acronym of organization.", required=True)
        name = graphene.String(description="Full name of organization.", required=True)
        zone = graphene.String(
            description="The zone which the organization belongs to.", required=True
        )
        sector = graphene.String(
            description="The sector which the organization belongs to.", required=True
        )
        province = graphene.String(
            description="The province in which the organization is located in.",
            required=True,
        )
        city = graphene.String(
            description="The city in which the organization is located in.",
            required=True,
        )

    # If the update passed or failed
    status = graphene.Boolean()

    @require_token
    def mutate(self, info, **kwargs):
        user_id = kwargs.get("user_id")
        user_roles = kwargs.get("user_roles")
        name = cleanse_input(kwargs.get("name"))
        acronym = cleanse_input(kwargs.get("acronym"))
        zone = cleanse_input(kwargs.get("zone"))
        sector = cleanse_input(kwargs.get("sector"))
        province = cleanse_input(kwargs.get("province"))
        city = cleanse_input(kwargs.get("city"))

        if is_super_admin(user_roles=user_roles):
            # Check to see if organization already exists
            slug = slugify_value(name)
            org_orm = (
                db_session.query(Organizations)
                .filter(Organizations.slug == slug)
                .first()
            )

            if org_orm is not None:
                logger.warning(
                    f"Super admin: {user_id} tried to create an organization, but it already exists."
                )
                raise GraphQLError("Error, unable to create organization.")

            # Generate org tags
            org_tags = {
                "zone": zone,
                "sector": sector,
                "province": province,
                "city": city,
            }

            # Create new org entry in db
            new_org = Organizations(name=name, acronym=acronym, org_tags=org_tags)

            # Add new org entry into the session
            db_session.add(new_org)

            # Push update to db and return status
            try:
                db_session.commit()
                logger.info(
                    f"Super admin: {user_id} successfully created a new org: {new_org.slug}."
                )
                return CreateOrganization(status=True)
            except Exception as e:
                db_session.rollback()
                db_session.flush()
                logger.error(
                    f"Database error occured when {user_id} tried to create a new organization {str(e)}"
                )
                raise GraphQLError("Error, unable to create organization.")
        else:
            logger.warning(
                f"User: {user_id} tried to create a new organization, but is not a super admin."
            )
            raise GraphQLError("Error, unable to create organization.")

File: api-js/src/mutations/organizations/create-organization.js
const { GraphQLNonNull, GraphQLString } = require('graphql')
const { mutationWithClientMutationId } = require('graphql-relay')
const { t } = require('@lingui/macro')
const { Acronym } = require('../../scalars')
const { organizationType } = require('../../types')

const createOrganization = new mutationWithClientMutationId({
  name: 'CreateOrganization',
  description:
    'This mutation allows the creation of an organization inside the database.',
  inputFields: () => ({
    acronymEN: {
      type: GraphQLNonNull(Acronym),
      description: 'The English acronym of the organization.',
    },
    acronymFR: {
      type: GraphQLNonNull(Acronym),
      description: 'The French acronym of the organization.',
    },
    nameEN: {
      type: GraphQLNonNull(GraphQLString),
      description: 'The English name of the organization.',
    },
    nameFR: {
      type: GraphQLNonNull(GraphQLString),
      description: 'The French name of the organization.',
    },
    zoneEN: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the zone the organization belongs to.',
    },
    zoneFR: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the zone the organization belongs to.',
    },
    sectorEN: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the sector the organization belongs to.',
    },
    sectorFR: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The French translation of the sector the organization belongs to.',
    },
    countryEN: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the country the organization resides in.',
    },
    countryFR: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The French translation of the country the organization resides in.',
    },
    provinceEN: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the province the organization resides in.',
    },
    provinceFR: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The French translation of the province the organization resides in.',
    },
    cityEN: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The English translation of the city the organization resides in.',
    },
    cityFR: {
      type: GraphQLNonNull(GraphQLString),
      description:
        'The French translation of the city the organization resides in.',
    },
  }),
  outputFields: () => ({
    organization: {
      type: organizationType,
      description: 'The newly created organization.',
      resolve: async (payload) => {
        return payload.organization
      },
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      request,
      collections,
      transaction,
      query,
      userId,
      auth: { userRequired },
      loaders: { orgLoaderBySlug },
      validators: { cleanseInput, slugify },
    },
  ) => {
    // Cleanse Input
    const acronymEN = cleanseInput(args.acronymEN)
    const acronymFR = cleanseInput(args.acronymFR)
    const nameEN = cleanseInput(args.nameEN)
    const nameFR = cleanseInput(args.nameFR)
    const zoneEN = cleanseInput(args.zoneEN)
    const zoneFR = cleanseInput(args.zoneFR)
    const sectorEN = cleanseInput(args.sectorEN)
    const sectorFR = cleanseInput(args.sectorFR)
    const countryEN = cleanseInput(args.countryEN)
    const countryFR = cleanseInput(args.countryFR)
    const provinceEN = cleanseInput(args.provinceEN)
    const provinceFR = cleanseInput(args.provinceFR)
    const cityEN = cleanseInput(args.cityEN)
    const cityFR = cleanseInput(args.cityFR)

    // Create EN and FR slugs
    const slugEN = slugify(nameEN)
    const slugFR = slugify(nameFR)

    // Get user
    const user = await userRequired()

    // Check to see if org already exists
    const [orgEN, orgFR] = await orgLoaderBySlug.loadMany([slugEN, slugFR])

    if (typeof orgEN !== 'undefined' || typeof orgFR !== 'undefined') {
      console.warn(
        `User: ${userId} attempted to create an organization that already exists: ${slugEN}`,
      )
      throw new Error(
        i18n._(t`Unable to create organization. Please try again.`),
      )
    }

    // Create new organization
    const organizationDetails = {
      verified: false,
      orgDetails: {
        en: {
          slug: slugEN,
          acronym: acronymEN,
          name: nameEN,
          zone: zoneEN,
          sector: sectorEN,
          country: countryEN,
          province: provinceEN,
          city: cityEN,
        },
        fr: {
          slug: slugFR,
          acronym: acronymFR,
          name: nameFR,
          zone: zoneFR,
          sector: sectorFR,
          country: countryFR,
          province: provinceFR,
          city: cityFR,
        },
      },
    }

    // Generate list of collections names
    const collectionStrings = []
    for (const property in collections) {
      collectionStrings.push(property.toString())
    }

    // Setup Trans action
    const trx = await transaction(collectionStrings)

    let cursor
    try {
      cursor = await trx.run(
        async () =>
          await query`
            INSERT ${organizationDetails} INTO organizations 
            RETURN MERGE({ _id: NEW._id, _key: NEW._key, _rev: NEW._rev, verified: NEW.verified }, TRANSLATE(${request.language}, NEW.orgDetails))
          `,
      )
    } catch (err) {
      console.error(
        `Transaction error occurred when user: ${userId} was creating new organization ${slugEN}: ${err}`,
      )
      throw new Error(
        i18n._(t`Unable to create organization. Please try again.`),
      )
    }
    const organization = await cursor.next()

    try {
      await trx.run(
        async () =>
          await query`
            INSERT {
              _from: ${organization._id},
              _to: ${user._id},
              permission: "admin"
            } INTO affiliations
          `,
      )
    } catch (err) {
      console.error(
        `Transaction error occurred when inserting edge definition for user: ${userId} to ${slugEN}: ${err}`,
      )
      throw new Error(
        i18n._(t`Unable to create organization. Please try again.`),
      )
    }

    try {
      await trx.commit()
    } catch (err) {
      console.error(
        `Transaction error occurred when committing new organization: ${slugEN} for user: ${userId} to db: ${err}`,
      )
      throw new Error(
        i18n._(t`Unable to create organization. Please try again.`),
      )
    }

    console.info(
      `User: ${userId} successfully created a new organization: ${slugEN}`,
    )

    return {
      organization: {
        id: organization._key,
        ...organization,
      },
    }
  },
})

module.exports = {
  createOrganization,
}
